// framer-motion/dist/es/animation/use-animated-state.js
import { __rest, __assign, __read } from 'tslib';
import { axisBox } from '../utils/geometry/index.js';
import { getOrigin, checkTargetForNewValues } from '../render/utils/setters.js';
import { animateVisualElement } from '../render/utils/animation.js';
import { visualElement } from '../render/index.js';
import { useState, useEffect } from 'react';
import { useConstant } from '../utils/use-constant.js';

var stateVisualElement = visualElement({
    createRenderState: function () { return ({}); },
    build: function () { },
    measureViewportBox: axisBox,
    resetTransform: function () { },
    restoreTransform: function () { },
    removeValueFromMutableState: function () { },
    render: function () { },
    scrapeMotionValuesFromProps: function () {
        return {};
    },
    readValueFromInstance: function (_state, key, options) {
        return options.initialState[key] || 0;
    },
    makeTargetAnimatable: function (element, _a) {
        var transition = _a.transition, transitionEnd = _a.transitionEnd, target = __rest(_a, ["transition", "transitionEnd"]);
        var origin = getOrigin(target, transition || {}, element);
        checkTargetForNewValues(element, target, origin);
        return __assign({ transition: transition, transitionEnd: transitionEnd }, target);
    },
});
/**
 * This is not an officially supported API and may be removed
 * on any version.
 * @internal
 */
function useAnimatedState(initialState) {
    var _a = __read(useState(initialState), 2), animationState = _a[0], setAnimationState = _a[1];
    var element = useConstant(function () {
        return stateVisualElement({ props: {} }, { initialState: initialState });
    });
    useEffect(function () {
        element.ref({});
        return function () { return element.ref(null); };
    }, []);
    useEffect(function () {
        element.setProps({
            onUpdate: function (v) { return setAnimationState(__assign({}, v)); },
        });
    });
    var startAnimation = useConstant(function () { return function (animationDefinition) {
        return animateVisualElement(element, animationDefinition);
    }; });
    return [animationState, startAnimation];
}

export { useAnimatedState };
